ledger/acctonline.go

```diff
@@ -619,99 +620,102 @@ func (ao *onlineAccounts) lookupOnlineAccountData(rnd basics.Round, addr basics.
+			}
+			// the account appears in the deltas, but we don't know if it appears in the
+			// delta range of [0..offset], so we'll need to check :
+			// Traverse the deltas backwards to ensure that later updates take
+			// priority if present.
+			for offset > 0 {
+				offset--
+				d, ok := ao.deltas[offset].GetData(addr)
+				if ok {
+					return d.OnlineAccountData(rewardsProto, rewardsLevel), nil
 				}
 			}
 		}
+	}
 
-		if macct, has := ao.onlineAccountsCache.read(addr, rnd); has {
-			return macct.GetOnlineAccountData(rewardsProto, rewardsLevel), nil
-		}
-
-		ao.accountsMu.RUnlock()
-		needUnlock = false
-
-		// No updates of this account in the in-memory deltas; use on-disk DB.
-		// The check in roundOffset() made sure the round is exactly the one
-		// present in the on-disk DB.  As an optimization, we avoid creating
-		// a separate transaction here, and directly use a prepared SQL query
-		// against the database.
-		persistedData, err = ao.accountsq.lookupOnline(addr, rnd)
-		if persistedData.round == currentDbRound {
-			if persistedData.rowid != 0 {
-				// if we read actual data return it
-				ao.accountsMu.Lock()
-				ao.onlineAccountsCache.clear(addr)
-				if !ao.onlineAccountsCache.full() {
-					ao.accountsMu.Unlock()
-					persistedDataHistory, _ := ao.accountsq.lookupOnlineHistory(addr)
-					ao.accountsMu.Lock()
-					for _, data := range persistedDataHistory {
-						ao.onlineAccountsCache.writeFront(
-							data.addr,
-							cachedOnlineAccount{
-								baseOnlineAccountData: data.accountData,
-								updRound:              data.updRound,
-							})
+	if macct, has := ao.onlineAccountsCache.read(addr, rnd); has {
+		return macct.GetOnlineAccountData(rewardsProto, rewardsLevel), nil
+	}
+
+	ao.accountsMu.RUnlock()
+	needUnlock = false
+
+	// No updates of this account in the in-memory deltas; use on-disk DB.
+	// As an optimization, we avoid creating
+	// a separate transaction here, and directly use a prepared SQL query
+	// against the database.
+	persistedData, err = ao.accountsq.lookupOnline(addr, rnd)
+	if persistedData.rowid != 0 {
+		// if we read actual data return it
+		err = func() error {
+			ao.accountsMu.Lock()
+			defer ao.accountsMu.Unlock()
+
+			ao.onlineAccountsCache.clear(addr)
+			if !ao.onlineAccountsCache.full() {
+				// lookupOnlineHistory does not return the account db round because of the following observation:
+				// 1. ao.onlineAccountsCache update happens with ao.accountsMu taken here and in postCommit
+				// 2. If we started reading here
+				// 2.1 before commitRound => OK, read what is in DB and then add new entries in postCommit
+				// 2.2 after commitRound but before postCommit => OK, read full history, ignore update from postCommit in writeFront's updRound comparison
+				// 2.3 after postCommit => OK, postCommit does not add new entry with writeFrontIfExist, but here load the full history
+				persistedDataHistory, err := ao.accountsq.lookupOnlineHistory(addr)
```

Comment:
Are you sure it's ok to keep the lock here? It will prevent other lookups from finishing and prevent the tracker db from advancing.

Reply:
lets see if all these scenarios work if moving it up to pre-lock

Reply:
sounds like there is no difference if read in or outside the lock - the db data consistency is guaranteed by ACID and the cache update is under ao.accountsMu lock.